src/firecracker/src/api_server_adapter.rs

```diff
@@ -0,0 +1,323 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::path::PathBuf;
+use std::sync::mpsc::{channel, Receiver, RecvError, Sender, TryRecvError};
+use std::sync::{Arc, RwLock};
+use std::thread;
+
+use api_server::{ApiServer, VmmRequest, VmmResponse};
+use mmds::MMDS;
+use utils::eventfd::EventFd;
+use vmm::controller::VmmController;
+use vmm::resources::VmResources;
+use vmm::vmm_config;
+use vmm::vmm_config::instance_info::InstanceInfo;
+
+struct ApiAdapter {
+    api_event_fd: EventFd,
+    from_api: Receiver<VmmRequest>,
+    to_api: Sender<VmmResponse>,
+}
+
+impl ApiAdapter {
+    pub fn new(
+        api_event_fd: EventFd,
+        from_api: Receiver<VmmRequest>,
+        to_api: Sender<VmmResponse>,
+    ) -> Self {
+        ApiAdapter {
+            api_event_fd,
+            from_api,
+            to_api,
+        }
+    }
+
+    /// Builds and starts a microVM. It either uses a config json or receives API requests to
+    /// get the microVM configuration.
+    ///
+    /// Returns a running `Vmm` object.
+    pub fn build_microvm_from_requests(
+        &self,
+        seccomp_level: u32,
+        // FIXME: epoll context can be polluted by failing boot attempts
+        epoll_context: &mut vmm::EpollContext,
+    ) -> (VmResources, vmm::Vmm) {
+        let mut vm_resources = VmResources::default();
+        // Configure and start microVM through successive API calls.
+        // Iterate through API calls to configure microVm.
+        // The loop breaks when a microVM is successfully started, and returns a running Vmm.
+        let mut built_vmm = None;
+        while built_vmm.is_none() {
+            match self.from_api.recv() {
+                Ok(vmm_request) => {
+                    // Also consume the API event. This is safe since communication
+                    // between this thread and the API thread is synchronous.
+                    let _ = self.api_event_fd.read().map_err(|e| {
+                        error!("VMM: Failed to read the API event_fd: {}", e);
+                        std::process::exit(i32::from(vmm::FC_EXIT_CODE_GENERIC_ERROR));
+                    });
+
+                    let (response, maybe_vmm) = self.handle_preboot_request(
+                        *vmm_request,
+                        &mut vm_resources,
+                        seccomp_level,
+                        epoll_context,
+                    );
+
+                    // Send back the result.
+                    self.to_api
+                        .send(response)
+                        .map_err(|_| ())
+                        .expect("one-shot channel closed");
+                    built_vmm = maybe_vmm;
```

Comment:
I do not personally like maybe_vmm. I would prefer that handle_preboot_requests would return a response type which contains an enum. Matching that enum to vmm built result should provide the loop exit signal.